Hmsg<-list(
  Load_help1=paste0("This app can read single cell RNA-seq data in several formats: <br>",
                    " - Individual files in plain text ('.txt', '.csv' with genes as rows and cells as columns), ",
                    "Hierarchical Data Format (.H5), or R's data format RDS (.RDS). <br>",
                    " - 10X Genomics output folder, or a folder containing the files 'barcodes.tsv.gz', ",
                    "'features.tsv.gz' and 'matrix.mtx.gz'.<br><br>"),
  Load_help2=paste0("Select an excel file or a plaint text table ('.csv', '.tsv', '.txt') with extra cell information. ",
                    "Cell IDs of the experiment must match the first column on the table. If there are cells in the ",
                    "dataset that are not included in the meta data table, a notification will pop up. <br>"),
  Load_help3=paste0("Differential expression results calculated in this application can be exported to RDS or Excel ",
                    "files. In order to continue with phases of the analysis that rely on differential expression ",
                    "results without having to compute them each time, like cell type imputation, you can reload ",
                    "them here."),
  
  QC_help1=paste0("<b>Grouping variable</b>: select a variable to group cells by, like 'cluster', 'sample', etc.<br>"),
  QC_help2=paste0("<b>Cell Cycle Scoring</b>: estimate the cell cycle phase of each cell based on canonical markers. [",
                  "<a href=https://www.science.org/doi/abs/10.1126/science.aad0501 ",
                  "style='color:rgb(23,59,99)'>Marker source</a>]<br>"),
  
  Filter_help1=paste0("<b>Additional filters</b>: set additional rules to filter cells: keep or remove cells ",
                      "by their values for the selected variables. <br>"),
  Filter_help2=paste0("<b>Regressing variables</b>: try to eliminate the effect of a variable on the dataset. ",
                      "Seurat will model the relationship between each gene's expression and the value of the given ",
                      "variable, generating a corrected expression matrix. Some variables commonly regressed could be ",
                      "cell cycle phase o percentage of mitochondrial RNA. <br>"),
  Filter_help3=paste0("<b>Normalization options</b>: pre-processing expression data to remove technical variability ",
                      "while preserving biological differences, so the variance of a gene reflects biological ",
                      "heterogeneity. <br> - 'LogNormalize' will scale UMI counts by sequencing depth, add pseudocounts, ",
                      "and log-transform the data. <br> - 'SCTransfrom' builds a generalized linear model for each gene, ",
                      "gathering information from genes with similar abundances to avoid overfitting. This approach is ","
                      more computationally intensive, but it's more effective, allowing to use more features in ",
                      "downstream analysis. <br>"),
  
  Merge_help1=paste0( "<b>Merge type</b>: <br> - 'Simple merge' will combine the raw count matrices of the experiments, ",
                      "without performing any correction. <br>- 'Sample integration' will try to correct for technical ",
                      "differences (such as batch effects) by identifying sets of cells (anchors) that are in a ",
                      "similar state across datasets. <br>" ),
  Merge_help2=paste0("<b>Normalization options</b>: pre-processing expression data to remove technical variability ",
                      "while preserving biological differences, so the variance of a gene reflects biological ",
                      "heterogeneity. <br> - 'LogNormalize' will scale UMI counts by sequencing depth, add pseudocounts, ",
                      "and log-transform the data. <br> - 'SCTransfrom' builds a generalized linear model for each gene, ",
                      "gathering information from genes with similar abundances to avoid overfitting. This approach is ","
                      more computationally intensive, but it's more effective, allowing to use more features in ",
                      "downstream analysis. <br>"),
  
  ExCl_help1=paste0("<b>Principal Component Analysis</b>: this section will help to determine the optimal amount of ",
                    "information needed to proceed with the analysis, having the goal of retaining most relevant ",
                    "biological information while keeping computation as light as possible."),
  ExCl_help2=paste0("<b>Minimum drop in standard deviation</b>: the optimal number of PCs to use is typically determined ",
                    "as the point at which PC's standard deviation stops decreasing. By default, the minimum drop in ",
                    "standard deviation is set at 5% for three consecutive principal components. <br>"),
  ExCl_help3=paste0("<b>Margin</b>: evaluate the effects on cell clustering of the number of PCs used, examining a range of ",
                    "optimal PCs ± margin (default is ±5). <br>"),
  ExCl_help4=paste0("<b>Clustering</b>: check the influence of clustering resolution in grouping cells. This application ",
                    "uses the SNN modularity optimization based clustering algorithm implemented in the package Seurat, ",
                    "in which a relevant parameter is 'resolution'. Clustering resolution will determine the ",
                    "<i>granularity</i> of the resulting cell groups, with low resolutions returning large and general ",
                    "clusters, and higher resolutions producing the opposite."),
  ExCl_help5=paste0("The optimization phase of this clustering algorithm has a stochastic component, which means that, ",
                    "from the same data, we could get slightly different results each time the cells are clustered. The ",
                    "specific result we get each time is determined by the 'seed' used, usually a random number. ",
                    "Using a set of seeds and comparing the results is possible to examine how much variation there ",
                    "is among clustering results, which tells us how ambiguous or <i>uncertain</i> the classification ",
                    "of each cell is. <br>"),
  ExCl_help6=paste0("Create 2D visualizations of the dataset colored by cluster for some of the resolutions tested.<br>"),
  
  Cluster_help1=paste0("Generate a definitive clustering for the cells in the dataset. <br>Clustering parameters can be ",
                       "chosen manually or determined automatically based on the dataset size"),
  Cluster_help2=paste0("The optimization phase of this clustering algorithm has a stochastic component, which means that, ",
                    "from the same data, we could get slightly different results each time the cells are clustered. The ",
                    "specific result we get each time is determined by the 'seed' used, ussualy a random number. ",
                    "Using a set of seeds and comparing the results is possible to examine how much variation there ",
                    "is among clustering results, which tells us how ambiguous or <i>uncertain</i> the classification ",
                    "of each cell is. <br>"),
  
  Visu_help1=paste0("As it happens with clustering, dimensional reduction techniques like UMAP and tSNE have an stochastic ",
                    "component as well. <br>Selecting 'Explore options' will generate a collection of several possible reductions ",
                    "for the dataset using different random seeds or testing a range of values for the algoritm's parameters ,",
                    "<br>The option 'Generate an embedding for the dataset' will let you choose the more relevant parameters for ",
                    "UMAP or tSNE reduction algorithms and store the cell coordinates in the dataset -although this coordinates ",
                    "can also be exported to a CSV file ",
                    "independently-. <br>"),
  Visu_help2=paste0("<b>Reduction Algorithms</b>: tSNE and UMAP are the most widely used dimensionality reduction algorithms, ",
                    "and each algorithm produces slightly different results depending on each dataset's particular ",
                    "characteristics -how many cells types we have, how large is difference between those cell types, how are ",
                    "the cell types related to each other...-. <br><br>Here are two brief guides on how to interpret ",
                    "<a href=https://distill.pub/2016/misread-tsne/ style='color:rgb(23,59,99)'>tSNE</a> and ",
                    "<a href=https://pair-code.github.io/understanding-umap/ style='color:rgb(23,59,99)'>UMAP</a> ",
                    "reductions and what are their strengths and shortcomings.<br>"),
  Visu_help3=paste0("<b>Reduction parameters</b>: define the most influential hyperparameters in each algorithm: <br><br>",
                    "   - <b>Perplexity</b> (tSNE) and <b>Number of neighbors</b> (UMAP): This two variables effectively control ",
                    "how many neighboring cells are taken into account when constructing the embedding. On one hand, large numbers ",
                    "will prioritize global relationships between cell groups but might lose fine detail. On the other hand, small ",
                    "values of 'perplexity' or 'number of neighbors' will put more emphasis on the local structure.<br><br>",
                    "   - <b>Minimum distance</b> (UMAP): this parameter controls how tight the cells will 'clump' together in ",
                    "the final plot. Larger values will make the points more disperse and viceversa.<br>"),
  Visu_help4=paste0("Since these two algorithms have a stochastic component, several executions with the same data and parameters ",
                    "can produce slightly different results. The '<i>seed</i>' controls which particular result we draw out of all ",
                    "the possibilities. <br><br>Seed options are:<br>",
                    "   - '<b>Random sample</b>': will produce six different plots using six random numbers as seeds. <br>",
                    "   - '<b>Default value</b>': will use the default values for the algorithms seeds.<br>",
                    "   - '<b>Set manually</b>': if you want to reproduce a particular 2D visualization, set the value for the seed ",
                    "manually. <br>"),
  
  DiffExpr_help1=paste0("Estimate which genes have significantly different expression patterns between groups of cells. <br>",
                        "Cells in the dataset can be grouped by their value in one or two variables, for instance, 'sample' ",
                        "and 'condition'. <br><br>Differential expression results exported from this application can be ",
                        "reloaded later on the 'Load data' tab. <br>"),
  DiffExpr_help2=paste0("<b>Analysis modes</b>: Depending on what groups you want to compare to calculate differential ",
                        "expression, you will have several options: <br>- '1 VS rest': the gene expression of each group ",
                        "of cells will be compared to the rest of the dataset. <br> - '1 VS 1': groups will be compared pair ",
                        "by pair. This can take a long time and most of the 1 VS 1 comparisons can be irrelevant, so in ",
                        "order to speed the process up it's possible to choose two particular groups of cells to compare.",
                        "<br><br>When groups are formed using two variables, let's say 'treatment' and 'cluster', there is ",
                        "also the option to do a conditional comparison: <i>inside each cluster, compare treatment <b>A</b> ",
                        "and treatment <b>B</b></i>. <br>"),
  
  featPl_help1=paste0("Choose up to four genes to represent their expression values in the dataset. <br>",
                       "If a 2D visualization for the dataset has not been computed, violin type plots will be generated. ",
                       "Otherwise, you will have to option to project gene expression values onto the 2D UMAP or tSNE ",
                      "visualizations generated in the 2D visualizations tab. <br><br>Copy or save the plots by right-",
                      "clicking in them.<br>"),
  featPl_help2=paste0("Select a grouping variable (like 'cluster' or 'sample') to split the violin or bar plots by."),
  featPl_help3=paste0("Choose up to four cell metadata variables to represent their values in the dataset. <br>",
                      "If a 2D visualization for the dataset has not been computed, violin type plots will be generated. ",
                      "Otherwise, you will have to option to project gene expression values onto the 2D UMAP or tSNE ",
                      "visualizations generated in the 2D visualizations tab. <br><br>Copy or save the plots by right-",
                      "clicking in them.<br>"),
  
  cType_help1=paste0("Estimate cell type based on differentially expressed genes between groups of cells. <br><br> There are two ",
                     "modes of cell type imputation available: <br><br>- <b>Automatic annotation</b>: using the R package <i>",
                     "<b>scCATCH</i></b>. Choose a grouping variable (i.e.:'cluster') and scCATCH will calculate differential ",
                     "expression and compare the results to their database of markers for healthy and cancer tissues. <br>",
                     "scCATCH is available for human and murine cells. <br><br>- <b>Manual Annotation</b>: if differential ",
                     "expression has been previously calculated for the experiment, this section will allow the user to define ",
                     "markers of their choice for their cell types of interest. <br>"),
  cType_help2=paste0("<b>Automatic annotation with scCATCH</b> <br><br>Select a grouping variable, like 'cluster', and the ","
                     tissue type closer to your source material. <br><br>scCATCH's method to calculate differential expression is ",
                     "not the same as the one used internally in this app, so the results might differ slightly. <br><br> A ",
                     "notification will pop up if less than 10 tissue markers are present in your dataset. <br><be>For more ",
                     "information on how scCATCH works internally, check their ",
                     "<a href=https://www.sciencedirect.com/science/article/pii/S2589004220300663 style='color:rgb(23,59,99)'>",
                     "publication.</a> or their <a href=https://github.com/ZJUFanLab/scCATCH style='color:rgb(23,59,99)'>",
                     "GitHub repository</a>. <br>"),
  cType_help3=paste0("<b>Manual annotation with custom markers</b> <br><br>Introduce the names of genes expressed in your cell ",
                     "types of interest and click the 'Add' button. <br><br>As markers for cell types are added, a table display below ",
                     "will be updated. If any markers added are not present in the dataset, a notification will pop up.<br><br>",
                     "Cell groups will be assigned a cell type based on the differential expression of the markers selected by ",
                     "the user. <br>The recommended mode of differential expression analysis for cell type imputation is '1 VS rest'. <br>" ),
  
  paths_help1=paste0("<b>Search for altered metabolic pathways</b>: based on the results of differential expression it's possible to look ", 
                     "for affected pathways.<br>Using gene sets from <a href=https://www.gsea-msigdb.org/gsea/msigdb/collections.jsp#C2 ",
                     "style='color:rgb(23,59,99)'>MSigDB</a> that characterize a large number of metabolic patways, we can calculate",
                     "if there is a significant enrichment in differentially expressed genes."),
  paths_help2=paste0("Plot the DE state of genes (up/down regulation, no change, etc) expressed in the analyzed groups of cells that ",
                     "belong to particular pathways.<br>Select either: <br>- One differential expression result and several pathway gene ",
                     "sets.<br>- Several or all of the differential expression results and one pathway geneset<br>")
  )

Hmsg$Visu_help5 <- Hmsg$Visu_help2 # for 2D visualization's 2nd mode (store embedding in the Seurat object)
Hmsg$Visu_help6 <- Hmsg$Visu_help3
Hmsg$Visu_help7 <- Hmsg$Visu_help4

Hmsg$featPl_help4 <- Hmsg$featPl_help2




